api/oc_core_res.c

```diff
@@ -47,7 +49,7 @@ static oc_platform_info_t oc_platform_info;
 
 static bool announce_con_res = false;
 static int res_latency = 0;
-static size_t device_count = 0;
+static OC_ATOMIC_UINT32_T g_device_count = 0;
```

Comment:
why change the type?

Reply:
`size_t` is one of those annoying types whose size depends on the architecture and for atomic operations I must have known width before choosing the operation (at least on Windows). 

Changing the type requires the least amount of changes. There are other options (using `sizeof` to detect the true width of the variable; or not using atomic operations, but protecting the variable with a mutex.), but this seemed the best. Though I kept `size_t` as return value of `oc_core_get_num_devices`, so that's now inconsistent. That shouldn't be a problem, unless we support platforms where `size_t` is 16-bit, which would mean that the return value of `oc_core_get_num_devices` would get truncated for values larger than UINT16_T max (65535). If that's a relevant case for us then I suggest adding a check to `oc_core_add_new_device` so the value is never greater than SIZE_MAX.

Reply:
well, I do not think the approach of the atomic is the right one anyway, please use critical sections around the size_t variables.

Reply:
@WAvdBeek 
Using critical sections is more expensive than atomic operations for CPU and generally, atomic operations are used for counters like device_count.  Please, could you explain it more?